package/MDAnalysis/core/_get_readers.py

```diff
@@ -25,7 +25,6 @@
 import copy
 import inspect
 import mmtf
-import parmed as pmd
```

Comment:
Given that the standard library has well-designed mocking machinery, is there a reason not to use that approach? I think you can modify mocked objects/methods with pretty much any values you want so they are easy to distinguish from the "real thing."

```python
try:
    import parmed as pmd:
except ImportError:
    from unittest import mock
    # custom mocking modifications as needed
    sys.modules['pmd'] = mock.MagicMock()
    sys.modules['pmd.Structure'] = mock.MagicMock()

Reply:
I didn't know about this, thanks!

Reply:
@tylerjereddy I understand how mocking is super helpful for unit testing but I can't figure out how you would use it for type recognition without, at some point, inspecting the module or class name?

Reply:
What does the mocked object need to be able to do?

Reply:
Here it would ideally replace  `if isinstance(obj, pmd.Structure)` as the second argument (mock parmed structure).

Reply:
This does the trick for me locally, with full diff (to Richard's most recent commit) below the fold & full test suite passing locally:



```python
try:
    import parmed as pmd
except ImportError:
    from unittest import mock
    pmd = mock.MagicMock(Structure=type(None))
```
<details>

Full diff:

```diff
diff --git a/package/MDAnalysis/core/_get_readers.py b/package/MDAnalysis/core/_get_readers.py
index 79e65a4c2..0194eb3c1 100644
--- a/package/MDAnalysis/core/_get_readers.py
+++ b/package/MDAnalysis/core/_get_readers.py
@@ -25,6 +25,13 @@ from six import raise_from
 import copy
 import inspect
 import mmtf
+
+try:
+    import parmed as pmd
+except ImportError:
+    from unittest import mock
+    pmd = mock.MagicMock(Structure=type(None))
+
 import numpy as np
 from MDAnalysis.lib.util import isstream
 
@@ -32,15 +39,6 @@ from .. import _READERS, _PARSERS, _MULTIFRAME_WRITERS, _SINGLEFRAME_WRITERS, _C
 from ..lib import util
 
 
-def _is_parmed_object(thing):
-    module = inspect.getmodule(thing.__class__)
-
-    if module is None:
-        return False
-    else:
-        return module.__name__.startswith('parmed')
-
-
 def get_reader_for(filename, format=None):
     """Return the appropriate trajectory reader class for `filename`.
 
@@ -102,7 +100,7 @@ def get_reader_for(filename, format=None):
         elif isinstance(filename, mmtf.MMTFDecoder):
             # mmtf slurps mmtf object
             format = 'MMTF'
-        elif _is_parmed_object(filename):
+        elif isinstance(filename, pmd.Structure):
             format = 'PARMED'
         else:
             # else let the guessing begin!
@@ -247,7 +245,7 @@ def get_parser_for(filename, format=None):
     if format is None:
         if isinstance(filename, mmtf.MMTFDecoder):
             format = 'mmtf'
-        elif _is_parmed_object(filename):
+        elif isinstance(filename, pmd.Structure):
             format = 'PARMED'
         else:
             format = util.guess_format(filename)
diff --git a/testsuite/MDAnalysisTests/coordinates/test_parmed.py b/testsuite/MDAnalysisTests/coordinates/test_parmed.py
index 319f43e7a..f9c624cf9 100644
--- a/testsuite/MDAnalysisTests/coordinates/test_parmed.py
+++ b/testsuite/MDAnalysisTests/coordinates/test_parmed.py
@@ -46,17 +46,6 @@ from MDAnalysisTests.datafiles import (
 
 pmd = pytest.importorskip('parmed')
 
-from MDAnalysis.core._get_readers import _is_parmed_object
-
-
-@pytest.mark.parametrize('thing,reference', [
-    ('foo', False),
-    ([1,2,3], False),
-    (pmd.load_file(GRO), True),
-])
-def test_is_parmed_object(thing, reference):
-    assert _is_parmed_object(thing) == reference
-
 
 
 class TestParmEdReaderGRO:

```

</details>

That said, I'm not going to block on this---I do like that I don't need to add a new function/unit test with my approach since we are leaning on the std library.

Reply:
This was a cool way to deal with missing imports! I'm not sure it's relevant any more now we're doing format hints though (#2533)